`Lectures`: Fix broken pdf preview for lecture attachments and attachment units

marlonnienaber opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check all tasks by putting an x in the [ ] (don't: [x ], [ x], do: [x]). Remove not applicable tasks and do not leave them unchecked -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- Remove tasks that are not applicable for your PR. Please only put the PR into ready for review, if all relevant tasks are checked! -->
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable, and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).


#### Server
No adjustments were made to the server.


#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).


### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
This PR fixes two bug issues: #10629 and #10498

The problem is related to the library that is used to work with PDFs:
pdfjs-dist relies on a worker script to render pdf documents. This script is included in the dependency. To tell the the package where to find this script the pdfjs-dist.GlobalWorkerOptions.workerSrc property must be initialized with an URL pointing to the script. If this property is not set the package tries to infer the URL but this is not guaranteed to succeed. People report that the fallback did not work anymore after they upgraded from the 3.x versions to 4.x versions. I assume this is why our setup broke as well. In several issues of the PDF.JS project, maintainers therefore suggest to always set the workerSrc manually (see for example [here](https://github.com/mozilla/pdf.js/issues/10478)). For setting the URL there are basically two options:

1) use worker script hosted on a CDN
The README of the PDF.JS project mentions that the package is hosted publicly on several CDNs. We could point workerSrc to the according URL of a CDN.
\+ simple fix
\- requires manual update of the hardcoded URL whenever the package is updated (-> hard to maintain)
\- functionality depends on the CDN serving the file correctly


2) host the script as a static asset ourselves
\+ easier to maintain (script is copied from node_modules to webapp/content/script on postinstall -> no manual update required)
\- security risk? (see CodeQL warning)


### Description
<!-- Describe your changes in detail -->

I went for Option 2.  I added commands to the postinstall hook in package.json that first create the src/main/webapp/content/scripts directory and then copy the worker script into the new folder. This causes the script to be publicly served at /content/scripts/pdf.worker.min.mjs where pdfjs-dist can use it.

Finally I excluded the script from linting, adapted tests to work with the new setup and removed old logic.

### Steps for Testing
<!-- Please describe in detail how reviewers can test your changes. Make sure to take all related features and views into account! Below is an example that you can refine. -->
Prerequisites:
- 1 Instructor

1. Log in to Artemis
2. Create a Lecture with one file-upload unit and one attachment (you can use the attached 
[dummy.pdf](https://github.com/user-attachments/files/19762527/dummy.pdf) for both)
3. Navigate to course management of the course where you created the lecture
4. Click on lectures
5. Click on the units button of your lecture
6. Verify that you can download and view (button with eye icon) the PDF 
7. Go back to lectures
8. Click on the attachments button of your lecture
9. Verify that you can view (button with eye icon) the PDF

### Testserver States
You can manage test servers using [Helios](https://helios.aet.cit.tum.de/). Check environment statuses in the [environment list](https://helios.aet.cit.tum.de/repo/69562331/environment/list). To deploy to a test server, go to the [CI/CD](https://helios.aet.cit.tum.de/repo/69562331/ci-cd) page, find your PR or branch, and trigger the deployment.

### Review Progress
<!-- Each PR should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Performance Review
<!-- See [Large Course Setup](https://github.com/ls1intum/Artemis/tree/develop/supporting_scripts/course-scripts/quick-course-setup) for the automation script that handles large course setup. -->
#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files modified in this PR here. You can use `supporting_script/code-coverage/generate_code_cov_table/generate_code_cov_table.py` to automatically generate the coverage table from the corresponding artefacts of your branch (follow the ReadMe for setup details). -->
<!-- Alternatively you can execute the tests locally (see build.gradle and package.json) or look into the corresponding artefacts. -->
<!-- The line coverage must be above 90% for changes files, and you must use extensive and useful assertions for server tests and expect statements for client tests. -->
<!-- Note: Confirm in the last column that you have implemented extensive assertions for server tests and expect statements for client tests. -->
<!--       Remove rows with only trivial changes from the table. -->
<!--
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ExerciseService.java | 85% | ✅                           |
| programming-exercise.component.ts | 95% | ✅              |
-->
#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| pdf-preview-thumbnail-grid.component.ts | 99.26% | ✅ ❌ |
| pdf-preview.component.ts | 92.07% | ✅ ❌ |



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Chores**
  - Updated the post-installation process to automatically copy the PDF.js worker script into the appropriate directory, ensuring it is available for use in the application.
  - Adjusted ESLint configuration to exclude the PDF.js worker script from linting checks.
  - Added the PDF.js worker script to the `.gitignore` to prevent it from being tracked in version control.

- **Refactor**
  - Changed the way the PDF.js worker script is loaded by setting the worker source dynamically at runtime instead of using a static import.

- **Tests**
  - Improved test mocks for PDF.js worker configuration to better simulate and track worker source assignment in test environments.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

marlonnienaber commented:
I am no security expert, but I think the CodeQL warning represents no thread (please definitely have a look and correct me if I am wrong!). It is triggered because of the publicly hosted worker script. The code of the script is only used by pdfjs-dist and can not be altered before it is used by the package as far as I know. No other component of Artemis runs the code. The script is also no secret and publicly available.

florian-glombik commented on src/main/webapp/app/lecture/manage/pdf-preview/pdf-preview-thumbnail-grid/pdf-preview-thumbnail-grid.component.ts:
```diff
@@ -61,6 +60,10 @@ export class PdfPreviewThumbnailGridComponent implements OnChanges {
     protected readonly faEyeSlash = faEyeSlash;
     protected readonly faGripLines = faGripLines;
 
+    constructor() {
+        PDFJS.GlobalWorkerOptions.workerSrc = '/content/scripts/pdf.worker.min.mjs';
+    }
```
Could you add a comment here why we need this? I think this will be hard to understand out of the context of this PR

marlonnienaber replied:
added the comment

florian-glombik commented on src/main/webapp/content/scripts/pdf.worker.min.mjs:
I guess the file needs to be called exactly like this? 
At least to me, the name seems rather undescriptive, but I am not familiar with the PDF feature - so it might be better to keep the current name if that is a common convention

marlonnienaber replied:
I would say it is better to keep the original name. Like this people who have never worked with pdf.js can more easily find their way around (if you for example google pdf.worker.min.mjs you will find the right ressources)

florian-glombik reviewed (commented):
I think that the second approach is better for our case too 👍 

As we are serving this static file anyways, why don't we just fix the issue? 

If I understand the warning correctly, wouldn't we just need to find out where the replacement takes place and adjust the order? https://github.com/ls1intum/Artemis/security/code-scanning/743

I am referring to 
```
The encoding function correctly handles ampersand before the other characters. For example, the string me & "you" is encoded as me &amp; &quot;you&quot;, and the string &quot; is encoded as &amp;quot;.

The decoding function, however, incorrectly decodes &amp; into & before handling the other characters. So while it correctly decodes the first example above, it decodes the second example (&amp;quot;) to " (a single double quote), which is not correct.

Instead, the decoding function should decode the ampersand last:
```

marlonnienaber commented:
> you added the file to the git repo even though it's copied using `postinstall`. I suggest, you remove it again and add it to gitignore.
> 
> The solution is super hacky. We should definitely look for alternative solutions in the future.

I totally agree. I explored several solutions before this one. But they all in the end boiled down to serving the file as a static asset and I was not able to find any other solutions while keeping PDF.js. I think it is a poor design choice by the library to require a URL pointing to the worker. It might be worth looking into alternative packages that can do the same thing in a cleaner way.

marlonnienaber commented:
> you added the file to the git repo even though it's copied using `postinstall`. I suggest, you remove it again and add it to gitignore.
> 
> The solution is super hacky. We should definitely look for alternative solutions in the future.

I removed the file and added it to gitignore.

florian-glombik commented on src/main/webapp/app/lecture/manage/pdf-preview/pdf-preview-thumbnail-grid/pdf-preview-thumbnail-grid.component.ts:
```diff
@@ -61,6 +60,16 @@ export class PdfPreviewThumbnailGridComponent implements OnChanges {
     protected readonly faEyeSlash = faEyeSlash;
     protected readonly faGripLines = faGripLines;
 
+    /**
+     * PDF.js requires a URL pointing to a worker script to process PDFs.
+     * The worker script is included in the package but is required to be served separately as a static asset.
+     * In Artemis the postinstall lifecycle hook (see package.json) copies the script from the package to the
+     * publicly served contents folder.
+     */
```
I think it would be good to directly include a link to the repo, and if there is a thread / issue that discusses this topic I would link it with `is required to be served separately as a static asset.`, because that statement might change in the future. 

Besides that the comment looks clear to me 👍 

```suggestion
    /**
     * <a href="https://github.com/mozilla/pdf.js">PDF.js</a> requires a URL pointing to a worker script to process PDFs.
     * The worker script is included in the package but is required to be served separately as a static asset.
     * In Artemis the postinstall lifecycle hook (see package.json) copies the script from the package to the
     * publicly served contents folder.
     */
```

marlonnienaber replied:
done

florian-glombik reviewed (changes requested):
Please have another look if we can link more resources that already discuss this topic / an open issue for it

=> So if there are changes and we don't need the workaround anymore this can be easily double checked

marlonnienaber commented:
> Is the underlying issue already reported to the maintainers of `PDF.js`? @marlonnienaber
> 
> At least on a first sight it seems like there is still some development going on? https://github.com/mozilla/pdf.js/graphs/contributors

Unfortunately we use [pdfjs-dist](https://github.com/mozilla/pdfjs-dist/graphs/contributors). The repo seems inactive.

florian-glombik commented:
> > Is the underlying issue already reported to the maintainers of `PDF.js`? @marlonnienaber
> > At least on a first sight it seems like there is still some development going on? https://github.com/mozilla/pdf.js/graphs/contributors
> 
> Unfortunately we use [pdfjs-dist](https://github.com/mozilla/pdfjs-dist/graphs/contributors). The repo seems inactive.

Yeah, that repo doesn't seem to be maintained anymore 😓
